src/CommunityToolkit.Maui/Services/PopupService.shared.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using CommunityToolkit.Maui.Core;
+using CommunityToolkit.Maui.Extensions;
+using CommunityToolkit.Maui.Views;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+
+namespace CommunityToolkit.Maui.Services;
+
+/// <inheritdoc cref="IPopupService"/>
+public class PopupService : IPopupService
+{
+	static readonly Dictionary<Type, Type> viewModelToViewMappings = [];
+
+	readonly IServiceProvider serviceProvider;
+
+	/// <summary>
+	/// Creates a new instance of <see cref="PopupService"/>.
+	/// </summary>
+	/// <param name="serviceProvider">The <see cref="IServiceProvider"/> implementation.</param>
+	[ActivatorUtilitiesConstructor]
+	public PopupService(IServiceProvider serviceProvider)
+	{
+		this.serviceProvider = serviceProvider;
+	}
+
+	/// <summary>
+	/// Creates a new instance of <see cref="PopupService"/>.
+	/// </summary>
+	public PopupService()
+	{
+		serviceProvider = IPlatformApplication.Current?.Services
+						  ?? throw new InvalidOperationException("Could not locate IServiceProvider");
+	}
+
+	internal static void AddPopup<
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupView>(
+		IServiceCollection services, ServiceLifetime lifetime)
+		where TPopupView : IView
+	{
+		viewModelToViewMappings.TryAdd(typeof(TPopupView), typeof(TPopupView));
+		Routing.RegisterRoute(typeof(TPopupView).FullName, typeof(TPopupView));
+
+		services.Add(new ServiceDescriptor(typeof(TPopupView), typeof(TPopupView), lifetime));
+	}
+
+	internal static void AddPopup<
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupView,
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupViewModel>(
+		IServiceCollection services, ServiceLifetime lifetime)
+		where TPopupView : IView
+		where TPopupViewModel : notnull
+	{
+		viewModelToViewMappings.TryAdd(typeof(TPopupViewModel), typeof(TPopupView));
+		Routing.RegisterRoute(typeof(TPopupViewModel).FullName, typeof(TPopupView));
+
+		services.TryAdd(new ServiceDescriptor(typeof(TPopupView), typeof(TPopupView), lifetime));
+		services.TryAdd(new ServiceDescriptor(typeof(TPopupViewModel), typeof(TPopupViewModel), lifetime));
+	}
+
+	internal static void AddPopup<
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupView,
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupViewModel>(
+		TPopupView popup, TPopupViewModel viewModel, IServiceCollection services, ServiceLifetime lifetime)
+		where TPopupView : IView
+		where TPopupViewModel : notnull
+	{
+		viewModelToViewMappings.TryAdd(typeof(TPopupViewModel), typeof(TPopupView));
+		Routing.RegisterRoute(typeof(TPopupViewModel).FullName, typeof(TPopupView));
```

Comment:
Do we want to register a route here for the user? I think it's wrong of us to both be registering a route in a method named `AddPopup` and to assume the user wants a route registered to the route name `typeof(TPopupViewModel).FullName`.

I see two options: 
1. (My preference) Remove this line of code 
2. Add `string routeName` as a parameter to allow the user to define the name of the route